core/domain/exp_domain.py

```diff
@@ -1925,12 +1928,112 @@ def _convert_states_v46_dict_to_v47_dict(cls, states_dict):
             state_domain.State.convert_html_fields_in_state(
                 state_dict,
                 html_validation_service.convert_svg_diagram_tags_to_image_tags)
+
+        return states_dict
+
+    @classmethod
+    def _convert_states_v47_dict_to_v48_dict(cls, exp_id, states_dict):
+        """Converts from version 47 to 48. Version 48 contains the attribute
+        image_sizes_in_bytes for subtitled_html.
+
+        Args:
+            exp_id: str. The id of the exploration to which are linked to.
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+
+        def get_image_sizes_in_bytes_dict_from_html(html_string):
+            """Calculate the image_sizes_in_bytes dict from an html string.
+
+            Args:
+                html_string: str. The html string from a SubtitledHtml object.
+
+            Returns:
+                dict. The mappings for each rich text image in the html_string
+                to its size in bytes.
+            """
+            image_filenames = (
+                html_cleaner.get_image_filenames_from_html_strings(
+                    [html_string]
+                ))
+            image_sizes_in_bytes = {}
+            for image_filename in image_filenames:
+                fs = fs_domain.AbstractFileSystem(
+                    fs_domain.GcsFileSystem(
+                        feconf.ENTITY_TYPE_EXPLORATION, exp_id))
+                image_size_in_bytes = len(fs.get(
+                    '%s/%s' % (
+                        constants.ASSET_TYPE_IMAGE, image_filename)))
+                image_sizes_in_bytes[image_filename] = image_size_in_bytes
+
+            return image_sizes_in_bytes
+
+        for state_dict in states_dict.values():
+
+            # For content.
+            html_string = state_dict['content']['html']
```

Comment:
This all seems a bit unnecessarily manual. Why don't you take a look at convert_html_fields_in_state() and see if it suits your needs? If so then the logic here can probably be simplified.

Reply:
convert_html_fields_in_state() is used to update the html fields of the states.
In our case, we are not updating the html of the states, but introducing a new attribute `image_size_in_bytes`, this cannot be achieved directly using the convert_html_fields_in_state() -- so, one option that I can think of is to write a similar function in html_validation_services.py, which is used to update the image_size_in_bytes dict of the states object.

The advantages of this would be that this function can be reused between exploration and question models. Secondly, it can also be used in the regular update flow for updating the image_size_in_bytes dict. WDYT ?

Reply:
Yup -- I agree, if you can't do it using the existing function then let's make a similar function and have convert_html_fields_in_state() use it too. Basically the logic for "traversing all the html parts of an exploration" should be in just one function in the entire codebase.

Thanks!